neurokit2/stats/cluster.py

```diff
@@ -375,6 +375,136 @@ def _cluster_mixture(data, n_clusters=2, bayesian=False, random_state=None, **kw
 
     return prediction, clusters, info
 
+#from sys import stdout
+#states, L = _frederic_aahc(data, n_clusters=3)
+
+def _frederic_aahc(data, n_clusters=3, doplot=False):
+    """Atomize and Agglomerative Hierarchical Clustering Algorithm
+    AAHC (Murray et al., Brain Topography, 2008)
+    Args:
+        data: EEG data to cluster, numpy.array (n_samples, n_channels)
+        N_clusters: desired number of clusters
+        doplot: boolean, plot maps
+    Returns:
+        maps: n_maps x n_channels (numpy.array)
+    """
+
+    def extract_row(A, k):
+        v = A[k,:]
+        A_ = np.vstack((A[:k,:],A[k+1:,:]))
+        return A_, v
+
+    def extract_item(A, k):
+        a = A[k]
+        A_ = A[:k] + A[k+1:]
+        return A_, a
+
+    def locmax(x):
+        """Get local maxima of 1D-array
+        Args:
+            x: numeric sequence
+        Returns:
+            m: list, 1D-indices of local maxima
+        """
+        dx = np.diff(x) # discrete 1st derivative
+        zc = np.diff(np.sign(dx)) # zero-crossings of dx
+        m = 1 + np.where(zc == -2)[0] # indices of local max.
+        return m
+
+    #print("\n\t--- AAHC ---")
+    nt, nch = data.shape
+
+    # --- get GFP peaks ---
```

Comment:
this should not be done in the clustering since its preprocesing

Reply:
The algorithm seems like it requires the data (before and after peak extraction) as well as the calculation of gfp to even do any clustering in the first place though